File: src/lib/forecast.ts
import { apiClient, withRetry } from "@/lib/api/client";

export type Horizon = 30 | 60 | 90;
export type Level = "aggregate" | "channel" | "campaignType" | "campaign";
export type Channel = "Google Ads" | "Meta Ads" | "Microsoft Ads";

export interface BudgetAllocation {
  google: number;
  meta: number;
  microsoft: number;
}

export interface ForecastPoint {
  day: number;
  date: string;
  p10: number;
  p25: number;
  p50: number;
  p75: number;
  p90: number;
  roas: number;
}

export interface ChannelForecast {
  name: Channel;
  spend: number;
  revenue: number;
  roas: number;
  contribution: number; // 0-1
  confidence: number; // 0-1
  trend: number[]; // sparkline
}

export interface CampaignTypeRow {
  type: string;
  spend: number;
  revenue: number;
  roas: number;
  ctr: number;
  conv: number;
}

export interface CampaignRow {
  id: string;
  name: string;
  channel: Channel;
  type: string;
  spend: number;
  revenue: number;
  roas: number;
  ctr: number;
  conv: number;
  confidence: number;
}

export interface ForecastResult {
  horizon: Horizon;
  totalBudget: number;
  revenue: { p10: number; p50: number; p90: number };
  roas: { p10: number; p50: number; p90: number };
  growth: number;
  confidence: number;
  timeline: ForecastPoint[];
  distribution: { bucket: number; density: number }[];
  channels: ChannelForecast[];
  campaignTypes: CampaignTypeRow[];
  campaigns: CampaignRow[];
  generatedAt: string;
}

export interface Insights {
  summary: string;
  drivers: string[];
  positives: string[];
  negatives: string[];
  seasonality: string;
  allocation: string;
  risks: string[];
  recommendations: string[];
  flags: string[];
}

/** Backend contract: POST /forecast { horizon, budget } -> ForecastResult
 * (see backend/src/forecastiq/api/routers/forecast.py and
 * backend/src/forecastiq/schemas/forecast.py — field names match this
 * interface exactly, so no mapping layer is needed here). */
export async function predictForecast(
  budget: BudgetAllocation,
  horizon: Horizon,
): Promise<ForecastResult> {
  return withRetry(async () => {
    const response = await apiClient.post<ForecastResult>("/forecast", { horizon, budget });
    return response.data;
  });
}

/** Backend contract: POST /insights { horizon, budget } -> Insights
 * (see backend/src/forecastiq/api/routers/insights.py). The AI never predicts
 * — it explains a forecast the backend recomputes server-side from the same
 * budget/horizon carried on `result` (never a client-supplied forecast blob).
 * Throws if GROQ_API_KEY isn't configured server-side (503); callers should
 * treat that as "insights unavailable", not retry it away. */
export async function generateInsights(result: ForecastResult): Promise<Insights> {
  const spendFor = (name: Channel) => result.channels.find((c) => c.name === name)?.spend ?? 0;
  const budget: BudgetAllocation = {
    google: spendFor("Google Ads"),
    meta: spendFor("Meta Ads"),
    microsoft: spendFor("Microsoft Ads"),
  };
  const response = await apiClient.post<Insights>("/insights", { horizon: result.horizon, budget });
  return response.data;
}

// Matches the real dataset's actual scale (recent 30-day run-rate spend per
// channel in backend/data/) — not an arbitrary illustrative number. The
// trained models are tree-based quantile regressors and do not extrapolate
// well far outside the spend levels they were trained on (this dataset's
// campaigns run in the thousands-to-low-hundred-thousands per month, not
// millions); a budget far above this will still return a forecast, but ROAS
// quality degrades the further the input strays from the training range.
export const INITIAL_BUDGET: BudgetAllocation = {
  google: 70_000,
  meta: 7_500,
  microsoft: 4_000,
};

File: backend/src/forecastiq/schemas/forecast.py
"""Pydantic mirrors of the TypeScript interfaces in src/lib/forecast.ts —
field-for-field, so the existing frontend components need zero changes once
Phase 9 wires them to these endpoints instead of the mock."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Horizon = Literal[30, 60, 90]


class BudgetAllocationSchema(BaseModel):
    google: float
    meta: float
    microsoft: float


class ForecastRequestSchema(BaseModel):
    horizon: Horizon
    budget: BudgetAllocationSchema


class RevenueBandSchema(BaseModel):
    p10: float
    p50: float
    p90: float


class ForecastPointSchema(BaseModel):
    day: int
    date: str
    p10: float
    p25: float
    p50: float
    p75: float
    p90: float
    roas: float


class DistributionPointSchema(BaseModel):
    bucket: float
    density: float


class ChannelForecastSchema(BaseModel):
    name: str
    spend: float
    revenue: float
    roas: float
    contribution: float
    confidence: float
    trend: list[float]


class CampaignTypeRowSchema(BaseModel):
    type: str
    spend: float
    revenue: float
    roas: float
    ctr: float
    conv: float


class CampaignRowSchema(BaseModel):
    id: str
    name: str
    channel: str
    type: str
    spend: float
    revenue: float
    roas: float
    ctr: float
    conv: float
    confidence: float


class ForecastResultSchema(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    horizon: int
    total_budget: float = Field(alias="totalBudget")
    revenue: RevenueBandSchema
    roas: RevenueBandSchema
    growth: float
    confidence: float
    timeline: list[ForecastPointSchema]
    distribution: list[DistributionPointSchema]
    channels: list[ChannelForecastSchema]
    campaign_types: list[CampaignTypeRowSchema] = Field(alias="campaignTypes")
    campaigns: list[CampaignRowSchema]
    generated_at: str = Field(alias="generatedAt")

    @classmethod
    def from_pipeline_output(cls, data: dict) -> "ForecastResultSchema":
        """`data` is exactly what `ForecastPipeline.forecast()` / `.simulate_budget()`
        returns — see forecastiq/models/forecasting.py."""
        return cls(
            horizon=data["horizon_days"],
            totalBudget=data["total_budget"],
            revenue=data["revenue"],
            roas=data["roas"],
            growth=data["growth"],
            confidence=data["confidence"],
            timeline=data["timeline"],
            distribution=data["distribution"],
            channels=data["channels"],
            campaignTypes=data["campaign_types"],
            campaigns=data["campaigns"],
            generatedAt=data["generated_at"],
        )


class ForecastRunSummarySchema(BaseModel):
    """Mirrors the `Row` interface in src/routes/forecast-history.tsx."""

    id: str
    date: str
    budget: float
    revenue: float
    roas: float
    confidence: float
    status: str

    @classmethod
    def from_orm_run(cls, run) -> "ForecastRunSummarySchema":
        total_budget = sum(run.budget.values()) if isinstance(run.budget, dict) else 0.0
        status = run.status.value if hasattr(run.status, "value") else str(run.status)
        return cls(
            id=f"F-{run.id}",
            date=run.created_at.isoformat(),
            budget=total_budget,
            revenue=run.revenue_p50,
            roas=run.roas_p50,
            confidence=run.confidence,
            status=status,
        )
